File: frontend/src/lib/compute/interpolation.ts
/**
 * Numerical utilities: interpolation, integration, linspace.
 */

/**
 * Log-log linear interpolation for stopping power data.
 * Returns a function that interpolates in log-log space.
 * Extrapolation is used at boundaries.
 */
export function makeLogLogInterpolator(
  energiesMeV: Float64Array,
  dedx: Float64Array,
): (energy: number | Float64Array) => number | Float64Array {
  const n = energiesMeV.length;
  const logE = new Float64Array(n);
  const logS = new Float64Array(n);
  for (let i = 0; i < n; i++) {
    logE[i] = Math.log(energiesMeV[i]);
    logS[i] = Math.log(dedx[i]);
  }

  function lookupScalar(energy: number): number {
    const le = Math.log(energy);

    // Binary search for interval
    let lo = 0;
    let hi = n - 1;

    if (le <= logE[0]) {
      // Extrapolate below
      if (n < 2) return Math.exp(logS[0]);
      const slope = (logS[1] - logS[0]) / (logE[1] - logE[0]);
      return Math.exp(logS[0] + slope * (le - logE[0]));
    }
    if (le >= logE[n - 1]) {
      // Extrapolate above
      if (n < 2) return Math.exp(logS[n - 1]);
      const slope =
        (logS[n - 1] - logS[n - 2]) / (logE[n - 1] - logE[n - 2]);
      return Math.exp(logS[n - 1] + slope * (le - logE[n - 1]));
    }

    // Binary search
    while (hi - lo > 1) {
      const mid = (lo + hi) >> 1;
      if (logE[mid] <= le) lo = mid;
      else hi = mid;
    }

    // Linear interpolation in log-log
    const t = (le - logE[lo]) / (logE[hi] - logE[lo]);
    return Math.exp(logS[lo] + t * (logS[hi] - logS[lo]));
  }

  return (energy: number | Float64Array): number | Float64Array => {
    if (typeof energy === "number") {
      return lookupScalar(energy);
    }
    const result = new Float64Array(energy.length);
    for (let i = 0; i < energy.length; i++) {
      result[i] = lookupScalar(energy[i]);
    }
    return result;
  };
}

/**
 * Trapezoidal integration of y over x.
 */
export function trapezoid(
  y: Float64Array,
  x: Float64Array,
): number {
  let sum = 0;
  for (let i = 1; i < x.length; i++) {
    sum += 0.5 * (y[i - 1] + y[i]) * (x[i] - x[i - 1]);
  }
  return sum;
}

/**
 * Linear interpolation (equivalent to np.interp).
 * x must be sorted ascending.
 * Values outside range get left/right fill values.
 */
export function interp(
  xNew: Float64Array,
  x: Float64Array,
  y: Float64Array,
  left: number = 0,
  right: number = 0,
): Float64Array {
  const result = new Float64Array(xNew.length);
  const n = x.length;

  for (let i = 0; i < xNew.length; i++) {
    const xv = xNew[i];
    if (xv <= x[0]) {
      result[i] = left;
    } else if (xv >= x[n - 1]) {
      result[i] = right;
    } else {
      // Binary search
      let lo = 0;
      let hi = n - 1;
      while (hi - lo > 1) {
        const mid = (lo + hi) >> 1;
        if (x[mid] <= xv) lo = mid;
        else hi = mid;
      }
      const t = (xv - x[lo]) / (x[hi] - x[lo]);
      result[i] = y[lo] + t * (y[hi] - y[lo]);
    }
  }

  return result;
}

/**
 * Generate evenly spaced values (like numpy.linspace).
 */
export function linspace(
  start: number,
  stop: number,
  num: number,
): Float64Array {
  const arr = new Float64Array(num);
  if (num === 1) {
    arr[0] = start;
    return arr;
  }
  const step = (stop - start) / (num - 1);
  for (let i = 0; i < num; i++) {
    arr[i] = start + i * step;
  }
  return arr;
}

File: core/src/interpolation.rs
//! Numerical utilities: interpolation, integration, linspace.

/// Generate evenly spaced values (like numpy.linspace).
pub fn linspace(start: f64, stop: f64, num: usize) -> Vec<f64> {
    if num == 0 {
        return Vec::new();
    }
    if num == 1 {
        return vec![start];
    }
    let step = (stop - start) / (num - 1) as f64;
    (0..num).map(|i| start + i as f64 * step).collect()
}

/// Trapezoidal integration of y over x.
pub fn trapezoid(y: &[f64], x: &[f64]) -> f64 {
    assert_eq!(y.len(), x.len());
    let mut sum = 0.0;
    for i in 1..x.len() {
        sum += 0.5 * (y[i - 1] + y[i]) * (x[i] - x[i - 1]);
    }
    sum
}

/// Linear interpolation (equivalent to np.interp).
/// `x` must be sorted ascending. Values outside range get left/right fill values.
pub fn interp(x_new: &[f64], x: &[f64], y: &[f64], left: f64, right: f64) -> Vec<f64> {
    let n = x.len();
    x_new
        .iter()
        .map(|&xv| {
            if xv <= x[0] {
                return left;
            }
            if xv >= x[n - 1] {
                return right;
            }
            // Binary search
            let mut lo = 0usize;
            let mut hi = n - 1;
            while hi - lo > 1 {
                let mid = (lo + hi) >> 1;
                if x[mid] <= xv {
                    lo = mid;
                } else {
                    hi = mid;
                }
            }
            let t = (xv - x[lo]) / (x[hi] - x[lo]);
            y[lo] + t * (y[hi] - y[lo])
        })
        .collect()
}

/// Log-log linear interpolation for stopping power data.
/// Returns a closure that interpolates in log-log space with extrapolation at boundaries.
pub fn make_log_log_interpolator(
    energies_mev: &[f64],
    dedx: &[f64],
) -> Box<dyn Fn(&[f64]) -> Vec<f64> + Send + Sync> {
    let n = energies_mev.len();
    let log_e: Vec<f64> = energies_mev.iter().map(|&e| e.ln()).collect();
    let log_s: Vec<f64> = dedx.iter().map(|&s| s.ln()).collect();

    Box::new(move |energies: &[f64]| {
        energies
            .iter()
            .map(|&energy| {
                let le = energy.ln();

                if le <= log_e[0] {
                    if n < 2 {
                        return log_s[0].exp();
                    }
                    let slope = (log_s[1] - log_s[0]) / (log_e[1] - log_e[0]);
                    return (log_s[0] + slope * (le - log_e[0])).exp();
                }
                if le >= log_e[n - 1] {
                    if n < 2 {
                        return log_s[n - 1].exp();
                    }
                    let slope = (log_s[n - 1] - log_s[n - 2]) / (log_e[n - 1] - log_e[n - 2]);
                    return (log_s[n - 1] + slope * (le - log_e[n - 1])).exp();
                }

                // Binary search
                let mut lo = 0usize;
                let mut hi = n - 1;
                while hi - lo > 1 {
                    let mid = (lo + hi) >> 1;
                    if log_e[mid] <= le {
                        lo = mid;
                    } else {
                        hi = mid;
                    }
                }

                let t = (le - log_e[lo]) / (log_e[hi] - log_e[lo]);
                (log_s[lo] + t * (log_s[hi] - log_s[lo])).exp()
            })
            .collect()
    })
}

/// Scalar version of log-log interpolation.
pub fn log_log_interp_scalar(energies_mev: &[f64], dedx: &[f64], energy: f64) -> f64 {
    let n = energies_mev.len();
    let le = energy.ln();

    let log_e_first = energies_mev[0].ln();
    let log_s_first = dedx[0].ln();

    if le <= log_e_first {
        if n < 2 {
            return log_s_first.exp();
        }
        let log_e_1 = energies_mev[1].ln();
        let log_s_1 = dedx[1].ln();
        let slope = (log_s_1 - log_s_first) / (log_e_1 - log_e_first);
        return (log_s_first + slope * (le - log_e_first)).exp();
    }

    let log_e_last = energies_mev[n - 1].ln();
    let log_s_last = dedx[n - 1].ln();

    if le >= log_e_last {
        if n < 2 {
            return log_s_last.exp();
        }
        let log_e_prev = energies_mev[n - 2].ln();
        let log_s_prev = dedx[n - 2].ln();
        let slope = (log_s_last - log_s_prev) / (log_e_last - log_e_prev);
        return (log_s_last + slope * (le - log_e_last)).exp();
    }

    // Binary search
    let mut lo = 0usize;
    let mut hi = n - 1;
    while hi - lo > 1 {
        let mid = (lo + hi) >> 1;
        if energies_mev[mid].ln() <= le {
            lo = mid;
        } else {
            hi = mid;
        }
    }

    let log_e_lo = energies_mev[lo].ln();
    let log_e_hi = energies_mev[hi].ln();
    let log_s_lo = dedx[lo].ln();
    let log_s_hi = dedx[hi].ln();
    let t = (le - log_e_lo) / (log_e_hi - log_e_lo);
    (log_s_lo + t * (log_s_hi - log_s_lo)).exp()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_linspace() {
        let v = linspace(0.0, 10.0, 5);
        assert_eq!(v.len(), 5);
        let expected = [0.0, 2.5, 5.0, 7.5, 10.0];
        for (a, b) in v.iter().zip(expected.iter()) {
            assert!((a - b).abs() < 1e-12);
        }
    }

    #[test]
    fn test_trapezoid_constant() {
        let y = vec![5.0, 5.0, 5.0, 5.0];
        let x = vec![0.0, 1.0, 2.0, 3.0];
        assert!((trapezoid(&y, &x) - 15.0).abs() < 1e-12);
    }

    #[test]
    fn test_log_log_power_law() {
        // y = x^2 -> log-log interpolation should be exact
        let x: Vec<f64> = vec![1.0, 2.0, 4.0, 8.0, 16.0];
        let y: Vec<f64> = x.iter().map(|&v| v * v).collect();
        let interp_fn = make_log_log_interpolator(&x, &y);
        let result = interp_fn(&[5.0]);
        assert!((result[0] - 25.0).abs() < 1e-6);
    }

    #[test]
    fn test_interp_basic() {
        let x = vec![0.0, 1.0, 2.0, 3.0];
        let y = vec![0.0, 10.0, 20.0, 30.0];
        let result = interp(&[0.5, 1.5, 2.5], &x, &y, 0.0, 0.0);
        assert!((result[0] - 5.0).abs() < 1e-12);
        assert!((result[1] - 15.0).abs() < 1e-12);
        assert!((result[2] - 25.0).abs() < 1e-12);
    }
}
